In [106]:
import sys
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from functools import reduce

### LECTURA DATOS

In [107]:
#mysql_config = {
#    'host': 'localhost',
#    'user': 'root',
#    'database': 'cancer_mama'
#}

#conn_mysql = mysql.connector.connect(**mysql_config)
#cursor_mysql = conn_mysql.cursor()

#query = "SELECT * FROM tumor;"
#df_sql = pd.read_sql_query(query, conn_mysql)

In [108]:
# PARA MAC
# df_sql = pd.read_csv('./Datos/datos_sql.csv')
# df_sql = df_sql.drop('Unnamed: 0',axis=1)
# df_sql

In [109]:
df_csv1 = pd.read_csv('./Datos/patients_batch1.csv')
df_csv2 = pd.read_csv('./Datos/patients_batch2.csv')
df_excel = pd.read_excel('./Datos/gynecological.xlsx')
df_excel = df_excel.drop('Unnamed: 0', axis=1)
df_json = pd.read_json('./Datos/histochemistry.json')
df_sql = pd.read_csv('./Datos/datos_sql.csv')
df_sql = df_sql.drop('Unnamed: 0', axis=1)

# PREPROCESADO

### Preprocesado gynecological


In [110]:
df_excel

ehr  pregnancy  birth  caesarean  abort  menarche_age  \
0     10011773       -8.0    0.0        NaN    0.0           NaN   
1     10030299        3.0    NaN        0.0    0.0           NaN   
2     10030824        0.0    0.0        0.0    NaN           NaN   
3     10053435        2.0    NaN        0.0    1.0          21.0   
4     10111454        0.0   -6.0        NaN    0.0          16.0   
...        ...        ...    ...        ...    ...           ...   
7681  99948591        9.0    2.0       -4.0    0.0          12.0   
7682  99961100        2.0    2.0        0.0    0.0          19.0   
7683  99981192        3.0    3.0        0.0   -5.0          11.0   
7684  99992350        4.0    3.0        NaN   -6.0          11.0   
7685  99994417        2.0   -6.0        NaN    NaN          13.0   

      menopause_age  
0              71.0  
1               NaN  
2              44.0  
3              74.0  
4              55.0  
...             ...  
7681           47.0  
7682           52.0  
7683            NaN  
7684           55.0  
7685            NaN  

[7686 rows x 7 columns]

In [111]:
df = df_excel

In [112]:
nulos = df.isnull()

num_nulos_por_columna = nulos.sum()

print("Número de valores nulos por columna:")
print(num_nulos_por_columna)


Número de valores nulos por columna:
ehr                 0
pregnancy        1273
birth            1269
caesarean        1555
abort            1451
menarche_age     2785
menopause_age    3200
dtype: int64


Imputamos los nulos de pregnancy, que es la suma de las otras 3 variables restantes, birth,caesarean y abort.

In [113]:
def imputar_pregnancy(row):
    if pd.isnull(row['pregnancy']):
        return row['birth'] + row['caesarean'] + row['abort']
    else:
        return row['pregnancy']

In [114]:
df['pregnancy'] = df.apply(imputar_pregnancy, axis=1)
df


ehr  pregnancy  birth  caesarean  abort  menarche_age  \
0     10011773       -8.0    0.0        NaN    0.0           NaN   
1     10030299        3.0    NaN        0.0    0.0           NaN   
2     10030824        0.0    0.0        0.0    NaN           NaN   
3     10053435        2.0    NaN        0.0    1.0          21.0   
4     10111454        0.0   -6.0        NaN    0.0          16.0   
...        ...        ...    ...        ...    ...           ...   
7681  99948591        9.0    2.0       -4.0    0.0          12.0   
7682  99961100        2.0    2.0        0.0    0.0          19.0   
7683  99981192        3.0    3.0        0.0   -5.0          11.0   
7684  99992350        4.0    3.0        NaN   -6.0          11.0   
7685  99994417        2.0   -6.0        NaN    NaN          13.0   

      menopause_age  
0              71.0  
1               NaN  
2              44.0  
3              74.0  
4              55.0  
...             ...  
7681           47.0  
7682           52.0  
7683            NaN  
7684           55.0  
7685            NaN  

[7686 rows x 7 columns]

In [115]:
nulos = df.isnull()

num_nulos_por_columna = nulos.sum()

print("Número de valores nulos por columna:")
print(num_nulos_por_columna)


Número de valores nulos por columna:
ehr                 0
pregnancy         442
birth            1269
caesarean        1555
abort            1451
menarche_age     2785
menopause_age    3200
dtype: int64


In [116]:
# Imputar valores nulos en la columna "birth"
def imputar_birth(row):
    if pd.isnull(row['birth']):
        return row['pregnancy'] - (row['caesarean'] + row['abort'])
    else:
        return row['birth']

# Aplicar la función a cada fila del DataFrame
df['birth'] = df.apply(imputar_birth, axis=1)

df

ehr  pregnancy  birth  caesarean  abort  menarche_age  \
0     10011773       -8.0    0.0        NaN    0.0           NaN   
1     10030299        3.0    3.0        0.0    0.0           NaN   
2     10030824        0.0    0.0        0.0    NaN           NaN   
3     10053435        2.0    1.0        0.0    1.0          21.0   
4     10111454        0.0   -6.0        NaN    0.0          16.0   
...        ...        ...    ...        ...    ...           ...   
7681  99948591        9.0    2.0       -4.0    0.0          12.0   
7682  99961100        2.0    2.0        0.0    0.0          19.0   
7683  99981192        3.0    3.0        0.0   -5.0          11.0   
7684  99992350        4.0    3.0        NaN   -6.0          11.0   
7685  99994417        2.0   -6.0        NaN    NaN          13.0   

      menopause_age  
0              71.0  
1               NaN  
2              44.0  
3              74.0  
4              55.0  
...             ...  
7681           47.0  
7682           52.0  
7683            NaN  
7684           55.0  
7685            NaN  

[7686 rows x 7 columns]

In [117]:
def imputar_caesarean(row):
    if pd.isnull(row['caesarean']):
        return row['pregnancy'] - (row['birth'] + row['abort'])
    else:
        return row['birth']

# Aplicar la función a cada fila del DataFrame
df['caesarean'] = df.apply(imputar_caesarean, axis=1)

df

ehr  pregnancy  birth  caesarean  abort  menarche_age  \
0     10011773       -8.0    0.0       -8.0    0.0           NaN   
1     10030299        3.0    3.0        3.0    0.0           NaN   
2     10030824        0.0    0.0        0.0    NaN           NaN   
3     10053435        2.0    1.0        1.0    1.0          21.0   
4     10111454        0.0   -6.0        6.0    0.0          16.0   
...        ...        ...    ...        ...    ...           ...   
7681  99948591        9.0    2.0        2.0    0.0          12.0   
7682  99961100        2.0    2.0        2.0    0.0          19.0   
7683  99981192        3.0    3.0        3.0   -5.0          11.0   
7684  99992350        4.0    3.0        7.0   -6.0          11.0   
7685  99994417        2.0   -6.0        NaN    NaN          13.0   

      menopause_age  
0              71.0  
1               NaN  
2              44.0  
3              74.0  
4              55.0  
...             ...  
7681           47.0  
7682           52.0  
7683            NaN  
7684           55.0  
7685            NaN  

[7686 rows x 7 columns]

In [118]:
def imputar_abort(row):
    if pd.isnull(row['abort']):
        return row['pregnancy'] - (row['caesarean'] + row['birth'])
    else:
        return row['abort']

# Aplicar la función a cada fila del DataFrame
df['abort'] = df.apply(imputar_abort, axis=1)

df

ehr  pregnancy  birth  caesarean  abort  menarche_age  \
0     10011773       -8.0    0.0       -8.0    0.0           NaN   
1     10030299        3.0    3.0        3.0    0.0           NaN   
2     10030824        0.0    0.0        0.0    0.0           NaN   
3     10053435        2.0    1.0        1.0    1.0          21.0   
4     10111454        0.0   -6.0        6.0    0.0          16.0   
...        ...        ...    ...        ...    ...           ...   
7681  99948591        9.0    2.0        2.0    0.0          12.0   
7682  99961100        2.0    2.0        2.0    0.0          19.0   
7683  99981192        3.0    3.0        3.0   -5.0          11.0   
7684  99992350        4.0    3.0        7.0   -6.0          11.0   
7685  99994417        2.0   -6.0        NaN    NaN          13.0   

      menopause_age  
0              71.0  
1               NaN  
2              44.0  
3              74.0  
4              55.0  
...             ...  
7681           47.0  
7682           52.0  
7683            NaN  
7684           55.0  
7685            NaN  

[7686 rows x 7 columns]

In [119]:
nulos = df.isnull()

num_nulos_por_columna = nulos.sum()

print("Número de valores nulos por columna:")
print(num_nulos_por_columna)

Número de valores nulos por columna:
ehr                 0
pregnancy         442
birth             447
caesarean         923
abort             670
menarche_age     2785
menopause_age    3200
dtype: int64


In [120]:
# Imputar valores negativos en "pregnancy" basándose en "birth" y "abort"
def imputar_pregnancy_negativo(row):
    if row['pregnancy'] < 0:
        return abs(row['birth'] + row['abort']+row['caesarean'])
    else:
        return row['pregnancy']

# Aplicar la función a cada fila del DataFrame
df['pregnancy'] = df.apply(imputar_pregnancy_negativo, axis=1)

# Mostrar el DataFrame después de la imputación
print("DataFrame después de imputar valores negativos en 'pregnancy':")
df

DataFrame después de imputar valores negativos en 'pregnancy':


ehr  pregnancy  birth  caesarean  abort  menarche_age  \
0     10011773        8.0    0.0       -8.0    0.0           NaN   
1     10030299        3.0    3.0        3.0    0.0           NaN   
2     10030824        0.0    0.0        0.0    0.0           NaN   
3     10053435        2.0    1.0        1.0    1.0          21.0   
4     10111454        0.0   -6.0        6.0    0.0          16.0   
...        ...        ...    ...        ...    ...           ...   
7681  99948591        9.0    2.0        2.0    0.0          12.0   
7682  99961100        2.0    2.0        2.0    0.0          19.0   
7683  99981192        3.0    3.0        3.0   -5.0          11.0   
7684  99992350        4.0    3.0        7.0   -6.0          11.0   
7685  99994417        2.0   -6.0        NaN    NaN          13.0   

      menopause_age  
0              71.0  
1               NaN  
2              44.0  
3              74.0  
4              55.0  
...             ...  
7681           47.0  
7682           52.0  
7683            NaN  
7684           55.0  
7685            NaN  

[7686 rows x 7 columns]

### Preprocesado patients_batch1/patients_batch2

Para el preprocesado vamos a comenzar uniendo los dataframes de origen csv, ya que comparten las mismas columnas. Para poder realizar el preprocesado de ambos de una manera más sencilla

In [121]:
df_csv = pd.merge(df_csv1, df_csv2, on=['ehr','birth_date','diagnosis_date','death_date'], how='outer')
df_csv

ehr  birth_date diagnosis_date death_date
0      10011773  1959-07-05     2015-04-02        NaN
1      10020495  1953-10-02     2017-12-04        NaN
2      10030299  1966-08-16     2019-06-27        NaN
3      10030824  1953-03-03     2018-09-07        NaN
4      10041592  1959-07-06     2018-11-19        NaN
...         ...         ...            ...        ...
10495  99966287  16/12/1978     02/08/2018        NaN
10496  99981192  10/04/1949     28/10/2017        NaN
10497  99988958  21/01/1966     27/05/2013        NaN
10498  99992350  04/09/1939     17/12/2018        NaN
10499  99994417  10/02/1962     01/04/2017        NaN

[10500 rows x 4 columns]

In [122]:
df_csv.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10500 entries, 0 to 10499
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   ehr             10500 non-null  int64 
 1   birth_date      10500 non-null  object
 2   diagnosis_date  10500 non-null  object
 3   death_date      1184 non-null   object
dtypes: int64(1), object(3)
memory usage: 410.2+ KB


Observamos que en nuestro dataset de csv tenemos variables de tipo int64 y object; El identificador del paciente tendrá valor numérico, luego el resto de variables (birth_date, diagnosis_date, death_date) con formato de objeto.

Tambien podemos observar gracaias a la columna Non-Null Count, que de las variables birth_date y diagnosis_date no tenemos ningun valor nulo (ya que tenemos 10500 valores no nulos). Sin embargo, en la columna death_date si que tenemos nulos que calcularemos más adelante

Primero vamos a pasar las fechas a tipo 'datetime'. Y una vez que los tenemos en este tipo, ponemos todas las fechas en el mismo formato

In [123]:
# Convertir las columnas objeto al formato datetime
df_csv['birth_date'] = pd.to_datetime(df_csv['birth_date'])
df_csv['diagnosis_date'] = pd.to_datetime(df_csv['diagnosis_date'])
df_csv['death_date'] = pd.to_datetime(df_csv['death_date'])

# Cambiamos los formatos a '%d/%m/%Y'
df_csv['birth_date'] = df_csv['birth_date'].dt.strftime('%d-%m-%Y')
df_csv['diagnosis_date'] = df_csv['diagnosis_date'].dt.strftime('%d-%m-%Y')
df_csv['death_date'] = df_csv['death_date'].dt.strftime('%d-%m-%Y')

df_csv

ehr  birth_date diagnosis_date death_date
0      10011773  05-07-1959     02-04-2015        NaN
1      10020495  02-10-1953     04-12-2017        NaN
2      10030299  16-08-1966     27-06-2019        NaN
3      10030824  03-03-1953     07-09-2018        NaN
4      10041592  06-07-1959     19-11-2018        NaN
...         ...         ...            ...        ...
10495  99966287  16-12-1978     08-02-2018        NaN
10496  99981192  04-10-1949     28-10-2017        NaN
10497  99988958  21-01-1966     27-05-2013        NaN
10498  99992350  09-04-1939     17-12-2018        NaN
10499  99994417  02-10-1962     04-01-2017        NaN

[10500 rows x 4 columns]

Como podemos apreciar a primera vista, y hemos observado antes,  la columna de death_date posee muchos nulos. Vamos a realizar un recuento de valores para esa columna para ver cuantos pacientes han fallecido de nuestro conjunto de datos

In [124]:
valores = df_csv['death_date'].count()
valores

1184

In [125]:
valores = df_csv['death_date'].isnull().sum()
valores

9316

Vemos que  hay 1184 pacientes que han fallecido, y 9316 pacientes estables.

Obseramos ahora si tenemos algún identificador duplicado, lo que significaría que tendríamos dos entradas del mismo paciente

In [126]:
df_csv['ehr'].duplicated().any()

True

Como nos devuelve True, tenemos al menos un identificador duplicado, vamos a imprimir cuantos

In [135]:
valores_duplicados = df_csv['ehr'].duplicated() # Función booleana
duplicados = df_csv[valores_duplicados]

# Pasamos los identificadores duplicados a formato lista para poder visualizarlos mejor
lista_duplicados = duplicados['ehr'].tolist()

# Imprimimos los valores duplicados, y cuantas veces aparecen duplicados
for i in lista_duplicados:
    conteo_duplicados = df_csv['ehr'].value_counts()[i]
    print(f"Valor: {i}, Contador: {conteo_duplicados}")

Valor: 72992494, Contador: 2
Valor: 73002338, Contador: 2
Valor: 73008149, Contador: 2
Valor: 73012939, Contador: 2
Valor: 73019870, Contador: 2
Valor: 73023533, Contador: 2
Valor: 73028729, Contador: 2
Valor: 73032048, Contador: 2
Valor: 73044771, Contador: 2
Valor: 73058633, Contador: 2
Valor: 73065270, Contador: 2
Valor: 73066125, Contador: 2
Valor: 73082648, Contador: 2
Valor: 73084641, Contador: 2
Valor: 73086658, Contador: 2
Valor: 73095852, Contador: 2
Valor: 73096542, Contador: 2
Valor: 73102549, Contador: 2
Valor: 73104918, Contador: 2
Valor: 73111704, Contador: 2
Valor: 73116349, Contador: 2
Valor: 73129838, Contador: 2
Valor: 73163695, Contador: 2
Valor: 73166702, Contador: 2
Valor: 73182879, Contador: 2
Valor: 73186443, Contador: 2
Valor: 73197959, Contador: 2
Valor: 73210647, Contador: 2
Valor: 73216753, Contador: 2
Valor: 73221718, Contador: 2
Valor: 73229306, Contador: 2
Valor: 73281640, Contador: 2
Valor: 73282540, Contador: 2
Valor: 73294146, Contador: 2
Valor: 7330744

In [141]:
fila_especificas = df_csv[df_csv['ehr'] == 77575810]
fila_especificas

ehr  birth_date diagnosis_date death_date
7495  77575810  21-03-1958     22-02-2014        NaN
7995  77575810  21-03-1958     22-02-2014        NaN

In [136]:
#df.drop_duplicates(subset='columna_a_verificar', keep='first', inplace=True)

Una vez que tenemos todos los pacientes correctamente, vamos a comprobar que las fechas de los pacientes entran en los rangos correctos:
- Dia: [1:31]
- Mes: [1:12]

Empezamos comprobando que el rango de los días es el correcto:

In [ ]:
columnas = ['birth_date', 'diagnosis_date', 'death_date']

for i in columnas:
    # Especifico el rango
    rango_días = (
        pd.to_datetime(df_csv[i], format='%d-%m-%Y', errors='coerce').dt.day < 1
        | (pd.to_datetime(df_csv[i], format='%d-%m-%Y', errors='coerce').dt.day > 31)
    )

    filas_fuera_de_rango = df_csv[rango_días]

    if not filas_fuera_de_rango.empty:
        print(f"Filas con días fuera del rango en {i}:")
        print(filas_fuera_de_rango)
    else:
        print(f"Todos los días en {i} están dentro del rango.")

Todos los días en birth_date están dentro del rango.
Todos los días en diagnosis_date están dentro del rango.
Todos los días en death_date están dentro del rango.


Comprobamos ahora el rango de los meses:

In [ ]:
columnas = ['birth_date', 'diagnosis_date', 'death_date']

for i in columnas:
    # Especifico el rango
    rango_mes = (
        pd.to_datetime(df_csv[i], format='%d-%m-%Y', errors='coerce').dt.month < 1
        | (pd.to_datetime(df_csv[i], format='%d-%m-%Y', errors='coerce').dt.month > 12)
    )

    filas_fuera_de_rango = df_csv[rango_mes]

    if not filas_fuera_de_rango.empty:
        print(f"Filas con días fuera del rango en {i}:")
        print(filas_fuera_de_rango)
    else:
        print(f"Todos los días en {i} están dentro del rango.")

Todos los días en birth_date están dentro del rango.
Todos los días en diagnosis_date están dentro del rango.
Todos los días en death_date están dentro del rango.


Por lo tanto, podemos concluir que tenemos todos los datos en el mismo formato, y con los valores correctos. Es decir, no tenemos ninguna fecha mal puesta

### Preprocesado histochemistry

In [ ]:
df_json

ehr   er  her2  ki67   pr
0     10011773  0.0   0.0  19.0  NaN
1     10020495  1.0   0.0   9.0  0.0
2     10030299  1.0   NaN  18.0  1.0
3     10030824  1.0   0.0   NaN  1.0
4     10041592  0.0   0.0  65.0  0.0
...        ...  ...   ...   ...  ...
9995  99966287  1.0   0.0   NaN  0.0
9996  99981192  1.0   0.0  14.0  0.0
9997  99988958  NaN   0.0  67.0  NaN
9998  99992350  0.0   0.0  16.0  0.0
9999  99994417  1.0   0.0  20.0  1.0

[10000 rows x 5 columns]

Lo pasamos a formato csv para poder estudiar los datos más facilmente

In [ ]:
df_json.to_csv('datos_json.csv',)
df_json

ehr   er  her2  ki67   pr
0     10011773  0.0   0.0  19.0  NaN
1     10020495  1.0   0.0   9.0  0.0
2     10030299  1.0   NaN  18.0  1.0
3     10030824  1.0   0.0   NaN  1.0
4     10041592  0.0   0.0  65.0  0.0
...        ...  ...   ...   ...  ...
9995  99966287  1.0   0.0   NaN  0.0
9996  99981192  1.0   0.0  14.0  0.0
9997  99988958  NaN   0.0  67.0  NaN
9998  99992350  0.0   0.0  16.0  0.0
9999  99994417  1.0   0.0  20.0  1.0

[10000 rows x 5 columns]

In [ ]:
df_json.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   ehr     10000 non-null  int64  
 1   er      9042 non-null   float64
 2   her2    9006 non-null   float64
 3   ki67    9032 non-null   float64
 4   pr      9018 non-null   float64
dtypes: float64(4), int64(1)
memory usage: 390.8 KB


Observamos que tenemos todos los datos de tipo numérico. El identificador del paciente (ehr) de tipo int, y el resto de variables como float

Mostramos ahora cuantos nulos hay por columna

In [ ]:
nulos = df_json.isnull()

num_nulos_por_columna = nulos.sum()

print("Número de valores nulos por columna:")
print(num_nulos_por_columna)

Número de valores nulos por columna:
ehr       0
er      958
her2    994
ki67    968
pr      982
dtype: int64


Tenemos varios valores 

### Preprocesado datos_sql.csv

In [ ]:
df_sql.to_csv('datos_sql.csv',)

In [ ]:
df_sql.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11167 entries, 0 to 11166
Data columns (total 14 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ehr                      11167 non-null  int64  
 1   n_tumor                  11167 non-null  int64  
 2   t_category               11167 non-null  object 
 3   n_category               11167 non-null  int64  
 4   m_category               11167 non-null  int64  
 5   t_category_after_neoadj  5677 non-null   object 
 6   n_category_after_neoadj  5677 non-null   float64
 7   m_category_after_neoadj  5627 non-null   float64
 8   stage_diagnosis          11167 non-null  object 
 9   stage_after_neo          6064 non-null   object 
 10  grade                    11167 non-null  int64  
 11  ductal                   6324 non-null   float64
 12  lobular                  1564 non-null   float64
 13  neoadjuvant              11167 non-null  object 
dtypes: float64(4), int64(5